crates/ruff/src/autofix/mod.rs

```diff
@@ -13,11 +13,20 @@ use crate::registry::{AsRule, Rule};
 pub(crate) mod codemods;
 pub(crate) mod edits;
 
+pub(crate) struct FixResult<'a> {
+    /// The resulting source code, after applying all fixes.
+    pub(crate) code: String,
+    /// The number of fixes applied for each [`Rule`].
+    pub(crate) fixes: FixTable,
+    /// The set of [`Edit`] that were applied.
+    pub(crate) edits: BTreeSet<&'a Edit>,
```

Comment:
Would it instead be possible to return a `SourceMap` that maps source positions to target positions in the `FixResult`. I would prefer that because it has a more narrow API and allows better re-use if we have other mapped types (e.g. python code in markdown files)

Reply:
I think it would roughly do the same as `update_cell_offsets` does today: 

The source map is defined as a `Vec<SourceMapMarker>`  where each marker stores the `source` to `destination` position:

https://github.com/charliermarsh/ruff/blob/d1c7a07d2ae56f4ff56b81c7f2a398e2262c42b9/crates/ruff_formatter/src/lib.rs#L285-L290

* Insertion: 
	* One marker at the start: `insertion_offset` -> `output.text_len()`
	* One marker at the end: `insertion_offset` -> `output.text_len()`
* Deletion:
	* One marker at the start: `deletion_start` -> `output.text_len()`
	* One marker at the end: `deletion_end` -> `output.text_len()`
* Replacement:
	* One marker at the start: `replacement_start` -> `output.text_len()`
	* One marker at the end: `replacement_end` -> `output.text_len()`


The entries are guaranteed to be sorted (at least per "apply fixes" iteration). We can now search the start and end of each cell (or use an iterator and make use of the sorted property to avoid `cell` count binary searches):

* cell start: Find the last source map entry that is smaller or equal to the cell start. The output position of the start is: `marker.dest + cell.start - marker.source` I think we need to use the first entry if multiple source marker start RIGHT at the start of the cell. 
* cell end: Find the last source map entry that ends before or at `cel.end`. You can map the end as: `marker.dest + marker.source - cell.end`

There are probably some corner cases that I didn't consider. 

One downside of this is that returning the `BTreeMap` is free because we build it anyway whereas building the source map has the cost of writing the source mark entries for each edit.